[WIP] Support for up to LINEAR_AXES 9

DerAndere1 opened the pull request:
### Description

This is early work in progress towards support for up to ten (EDIT: 9) non-extruder axes. No inverse kinematics are included. Like my previous 6-axis-support, the new code is Mostly framed by ´#if LINEAR_AXES >= ... #endif` preprocessor directives. 
 
Contributors: Paloky, HendrikJan-5D, DerAndere, thinkyhead

TODO:

- [x] Review and clean-up
- [x] Test compilation
- [x] Add sanity checks
- [x] quick hardware test with LINEAR_AXES 9
- [x] Thorough hardware test with LINEAR_AXES >= 7
- [ ] documentation for users

### Requirements

LINEAR_AXES 9

### Benefits

Extends multi-axis support to up to 10 (EDIT: 9) axes that benefit from endstops and auto homing. Axis names can be configured (Internal axis names are X,Y,Z,I,J,K,U,V,W). The extra axes are not so useful out of the box because inverse kinematics are missing, but this work can be used as the basis to develop multi axis machines. So far, the extra axes were used to control additional steppers that are not involved in positioning of the main tool.

### Configurations


[7axis_configs.zip](https://github.com/MarlinFirmware/Marlin/files/7407013/7axis_configs.zip)

[9axis_configs.zip](https://github.com/MarlinFirmware/Marlin/files/7407014/9axis_configs.zip)

### Related Issues

https://github.com/MarlinFirmware/Marlin/issues/22488

DerAndere1 committed:
Support for up to 10 non-extruder axes ( LINEAR_AXES 10 )

Co-authored-by: DerAndere <26200979+DerAndere1@users.noreply.github.com>
Co-authored-by: Paloky <paloky@gmail.com>

DerAndere1 committed:
Cleanup after addition of support for up to 10 linear axes

Thanks to HendrikJan-5D (https://github.com/HendrikJan-5D) for some of
this

ellensp commented:
this is getting absurd....

DerAndere1 commented on Marlin/src/pins/pins_postprocess.h:
```diff
@@ -1232,6 +1371,329 @@
+      #define AUTO_ASSIGNED_O_STEPPER 1
+    #endif
+  #endif
+  #if AXIS_HAS_SPI(K) && !defined(O_CS_PIN)
```
K should be O

DerAndere1 commented on Marlin/src/pins/pins_postprocess.h:
```diff
@@ -494,20 +534,119 @@
 // Filament Sensor first pin alias
-#if HAS_FILAMENT_SENSOR
-  #define FIL_RUNOUT1_PIN FIL_RUNOUT_PIN
+#if HAS_FILAMENT_SENSOK_  #define FIL_RUNOUT1_PIN FIL_RUNOUT_PIN
```
typo

DerAndere1 commented on Marlin/src/lcd/language/language_en.h:
```diff
@@ -276,6 +284,7 @@ namespace Language_en {
   LSTR MSG_INTENSITY_W                    = _UxGT("White Intensity");
   LSTR MSG_LED_BRIGHTNESS                 = _UxGT("Brightness");
 
+<<<<<<< Upstream, based on origin/bugfix-2.0.x
```
Merge confusion, see below

DerAndere1 commented on Marlin/src/gcode/parser.h:
```diff
@@ -225,7 +225,13 @@ class GCodeParser {
   #endif // !FASTER_GCODE_PARSER
 
   // Seen any axis parameter
+<<<<<<< Upstream, based on origin/bugfix-2.0.x
```
merge confusion

DerAndere1 commented on Marlin/src/gcode/feature/trinamic/M122.cpp:
```diff
@@ -52,13 +52,32 @@ void GcodeSuite::M122() {
         tmc_set_report_interval(MONITOR_DRIVER_STATUS_INTERVAL_MS);
     #endif
 
+<<<<<<< Upstream, based on origin/bugfix-2.0.x
```
Merge artifacts can be deleted

DerAndere1 commented on Marlin/src/gcode/config/M217.cpp:
```diff
@@ -33,6 +33,47 @@
 
 #include "../../MarlinCore.h" // for SP_X_STR, etc.
 
+<<<<<<< Upstream, based on origin/bugfix-2.0.x
```
Merge confusion

DerAndere1 commented on Marlin/src/feature/tmc_util.h:
```diff
@@ -335,14 +335,29 @@ void tmc_print_current(TMC &st) {
 void monitor_tmc_drivers();
+<<<<<<< Upstream, based on origin/bugfix-2.0.x
 void test_tmc_connection(LOGICAL_AXIS_DECL(const bool, true));
+=======
```
keep bugfix version here and below, delete other version

DerAndere1 reviewed (commented):
TODO: LINEAR_AXIS >=7 with SAFE_HOMING, IMPROVED_HOMING_RELIABILITY, SENSORLESS_HOMING, QUICK_HOMING

DerAndere1 commented:
I know, @ellensp  :D But I I obtained the code and wanted to present it here. I do not recommend to merge this, as the maintainability suffers, even with the latest refactorings.

thinkyhead committed:
Cleanup, fix

thinkyhead commented on Marlin/src/lcd/language/language_en.h:
```diff
@@ -78,6 +78,10 @@ namespace Language_en {
+  LSTR MSG_AUTO_HOME_M                    = _UxGT("Home ") LCD_STR_M;
+  LSTR MSG_AUTO_HOME_O                    = _UxGT("Home ") LCD_STR_O;
+  LSTR MSG_AUTO_HOME_P                    = _UxGT("Home ") LCD_STR_P;
+  LSTR MSG_AUTO_HOME_Q                    = _UxGT("Home ") LCD_STR_Q;
```
I think strings that contain axis names will need to get the `lcd_put_u8str_ind_P` treatment, so the `@` character will stand in for the axis name, similar to how `=`, `~`, and `*` are now used as placeholders for 1-based index, 0-based index, and indexed-E. That will reduce each set of 10 strings down to just one string.

DerAndere1 committed:
Support for G2, G3, M569 and M913 with LINEAR_AXES 10

DerAndere1 commented:
I added some more missing pieces. You can squash our fixup commits, if you like.

thinkyhead commented:
Oops, you seem to have introduced some typos. I have to go and sleep for a while, so please fix those and I will check in later.

thinkyhead committed:
Merge 'upstream/bugfix-2.0.x' into pr/22822

thinkyhead committed:
Cleanup, fix

thinkyhead commented on Marlin/src/module/endstops.cpp:
```diff
@@ -450,20 +534,24 @@ void Endstops::event_handler() {
-      #define P_AXIS Z_AXIS
-      if (TEST(hit_state, Z_MIN_PROBE)) _ENDSTOP_HIT_ECHO(P, 'P');
+      #define S_AXIS Z_AXIS
+      if (TEST(hit_state, Z_MIN_PROBE)) _ENDSTOP_HIT_ECHO(S, 'S');
```
'P' for Probe is sensible, and may be considered part of the protocol. Perhaps `TUVW` would be the better choice for axis names, rather than `MOPQ`. The `E` parameter is used in most instances to refer to the "Tool" so `T` should actually be non-conflicting in G-codes that take axis letters. Of course, with 10 axes the chances of parameter conflict is high.

DerAndere1 replied:
I would prefer to remove the 10th axis and go for `UVW`. I guess the confusion that may arise with configs that map the `I_AXIS` to letter U (`AXIS4_NAME 'U'`) would be acceptable.

thinkyhead replied:
Sure. I can live with only 9 linear axes.

DerAndere1 replied:
Thought about it once more. The name `U_AXIS` implies existance of secondary linear axes (usually parallel to X). This is nonsense for machines that have mainly rotational axes or that only have a linear axis parallel to z. Whenever possible, axis IDs `W` should stay reserved for references to a linear axis parallel to Z. Otherwise the implementation of the inverse kinematics for e.g. 7 axis robot arms or [for swiss style lathes with two independent z axes](https://cncmanual.com/download/2976/) will become very confusing. For now, I will remove the current P_AXIS and revert to `P` for probe.

DerAndere1 committed:
Fix extra axes stuck in absolute mode with EXTRUDERS > 1 (issue 22681)

DerAndere1 commented:
Commit https://github.com/MarlinFirmware/Marlin/pull/22822/commits/7ecc4c1f4c203fd1aa5f86c1b88f7dd4fbb92d98 is worth cherry picking into bugfix-2.0.x .

thinkyhead commented:
Thanks! Already done.

DerAndere1 committed:
Remove P axis, revert to P for Probe, fixes

DerAndere1 commented on Marlin/src/module/planner.cpp:
```diff
@@ -3124,14 +3092,14 @@ bool Planner::buffer_line(const xyze_pos_t &cart, const_feedRate_t fr_mm_s, cons
       const xyze_pos_t cart_dist_mm = LOGICAL_AXIS_ARRAY(
         cart.e - position_cart.e,
         cart.x - position_cart.x, cart.y - position_cart.y, cart.z - position_cart.z,
-        cart.i - position_cart.i, cart.j - position_cart.j, cart.j - position_cart.k,
```
This looks like a copy and paste mistake in bugfix-2.0.x causing a bug in the K axis:
https://github.com/MarlinFirmware/Marlin/blob/e85f6225053de257ee6f2652f097d0939d0dbde7/Marlin/src/module/planner.cpp#L3001
https://github.com/MarlinFirmware/Marlin/blob/e85f6225053de257ee6f2652f097d0939d0dbde7/Marlin/src/module/planner.cpp#L3006

DerAndere1 committed:
Towards SENSORLESS_HOMING with LINEAR_AXES >= 4

thinkyhead commented:
Incidentally, if the reason that "L" axis is being skipped is due to the use of "`l_axis`" in the arc planner, we can certainly change the naming of that in `plan_arc`. It just stands for "linear" but it only refers to _all axes not involved in the arc_, which are doing plain linear interpolation. The internal names we now use… IJKMOQ are simply internal, so of course we could call them `IJKLMN` … or go straight to `IJKUVW` … or maybe even `RSTUVW`. I tend to favor `IJKLMN` (even though both `M` and `N` are contraindicated for use due to Marlin's historical strictness about (almost always) avoiding the use of `G`, `M`, and `N` as parameter words) or `RSTUVW`.

In the far out future, it would be nice to see all the scattered axis-oriented elements (as illustrated by all the files we are now augmenting) get consolidated into an `Axis` class, which itself is built from some sort of `Motor` class, and so on. But when it comes to giant refactors (which would tend to make Marlin 32-bit-only) my general attitude is that RepRapFirmware or something like it would probably be a better starting-point than to rebuild Marlin from scratch.

DerAndere1 commented:
In terms of functionality, there would have been no conflict between `axis_l` in `plan_arc` and `L_AXIS`. Likewise, there is no interference between `axis_q` and `Q_AXIS`. The most important thing is to continue with restricting `AXIS*_NAME` to standard conforming letters XYZABCUVW. In most G/M-codes, those letters are used as parameter names to refer to a certain axis (the gcode parser is looking for `AXIS*_NAME`, not for hard-coded letters). For internal IDs, we can use for whatever you like, but I would prefer to do the renaming once only. `IJKUVW` is worth trying. `RSTUVW`would be a lot of work with not much gain:

Test case M907:
Letters `BTS` are reserved for existing parameters. 
This collides with `AXIS5_NAME 'B'`,  so we have to hard-code parameter names for referencing axes. It would make sense to use the internal axis IDs as user-facing parameter names in this special case.
`M907 X1 Y1 Z1 R1 S1 T1 U1 V1 W1 E1 B1 S1` (conflicting S. Also, T1 could be misinterpreted as tool selection)
`M907 X1 Y1 Z1 I1 J1 K1 M1 O1 Q1 E1 B1 S1` (M would be misinterpreded as second M-code)
`M907 X1 Y1 Z1 I1 J1 K1 U1 V1 W1 E1 B1 S1` (works)

Test case M217:
`ABEFGLQPRSUVWXYZ` parameter names are taken. 
leaves `CDHIJKO` as available parameter names  for extra axes. Again, using internal axis IDs works at least for axes IJK.

DerAndere1 committed:
Fix compilation with LINEAR_AXES 9

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/22822

thinkyhead committed:
post merge

thinkyhead committed:
Merge 'upstream/bugfix-2.0.x' into pr/22822

thinkyhead committed:
missing comma

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/22822

thinkyhead committed:
tmc_util tweak

DerAndere1 committed:
rename internal axis names MOQ -> UVW

DerAndere1 committed:
Support for LINEAR_AXES >= 4 for more M codes.

DerAndere1 committed:
auto pin assign warnings for LINEAR_AXES >= 7

DerAndere1 committed:
Fix renaming of axes MOQ to UVW

DerAndere1 commented:
Axis renaming to UVW is done. Also, SENSORLESS_HOMING, M217 and M907 now support LINEAR_AXES 9. 

Now, there is a request to add a new option like `#define SAFE_PROBING_POSITION { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 } `. At the beginning of the bed leveling procedure (G29), all axes should move to that position. Would it work if I simply add the following at the beginning of G29.cpp ?: 
```
xyz_pos_t safe_probing_position = SAFE_PROBING_POSITION;
current_position = safe_probing_position;
sync_plan_position();
```

DerAndere1 committed:
Fix compilation with LINEAR_AXES 7

DerAndere1 committed:
Add SAFE_BED_LEVELING_START_[XYZIJKUVW]

Settings to ensure that z probe is perpendicular
to bed at beginning of bed leveling procedure,
even with LINEAR_AXES >= 4.

DerAndere1 committed:
cleanup

DerAndere1 committed:
axes IJKUVW should not move to pos 0 when probing

DerAndere1 committed:
whitespace and revert variable names

DerAndere1 commented:
This was sucessfully tested by HendrikJan-5D in hardware, including a full continuous 7-axis print using a handwritten gcode script. I opened a new PR that superseeds this PR: https://github.com/MarlinFirmware/Marlin/pull/23112 (no new features, just a fresh rebase)
